### PR TITLE
Resolve Dtypes

### LINKED ISSUE #148: Consistent Class dtypes
While beginning to work with the code and make minor changes I found several areas (both new classes and Matlab ports) where numerically reasonable dtype handling did not exist or was inconsistent.  IE a lot of silent mixed mode.

For example, some classes do not have a dtype attribute, but clearly should. In other cases, a class may be instantiated as, or with, `single` precision, but internally creates arrays or invoke functions which use/return a default datatype which is `double`, without warning.  Some cases are setup to use doubles, but then instantiate classes that default in some internal way to `single`.

Besides being confusing, it creates problems that will become increasingly harder to find.

I propose the following:

- Classes that handle ndarray like things should have a dtype attribute
- This attribute should handle numpy defined dtypes, but may admit (with polite deprecation warning) 'single'/'double' to support old code.
- If a class is instantiated with or passed arrays, we should check their dtypes are consistent, and inherit that type if it is a constructor.
  - Otherwise default to double, unless there is strong proponent of singles.
- Any mixed mode encountered should generate a warning or raise error.
  - Halting is simpler.  To generate warnings we need to decide whether we want to upcast or downcast on mix. Both have concerning implications. By halting, maybe with a suggestion, we let the user decide.

### PR DESCRIPTION
**Pull request by garrettwrong**
Resolves #148 

Getting a head start on tightening up ASPIRE dtypes.  First commit (200 lines or so) seems to be enough to switch cov2d example between singles and doubles at top of file and actually get that precision at the numerical code.  Mainly this addresses `nufft` and `anufft`being passed inconsistent types for the 2d case.

Similar changes for 3d coming. Probably after that I'll make a list of the classes and have to more methodically check them. RIght now I'm focusing on the more striking type mismatches that come up in our tutorials/unit-tests.

This also addressed a circular dependence which has come up in another recent PR (`aspire.utils.misc` imports `Volume`, but methods in `misc` were needed by `Image`, thus needed by `Volume` and so on. I moved the obviously independent methods out...).

### CODE REVIEW

**Comment by garrettwrong** [comment]
This should cover implemented classes which contain or interact with numerical arrays, either by adding a `dtype` param or by inferring (usually from a `src` param).

Additionally this converts ASPIRE towards the float32 default decided in our recent team meeting.

Resolves many of the type mismatch warnings (virtually all, excluding ones tied to FINUFFT v1 which is hardcoded doubles).

Tutorials and unit tests have a `dtype` variable which can be changed to float64 (and actually changes the behavior e2e).

**Comment by garrettwrong** [comment]
Reviewing open issues, I think this will also resolve #149 . Most numerically sensitive unit tests should now have a dtype option that will run in either mode (defaulting to singles now). If there are any stragglers we can make small touch up issues.

File: src/aspire/estimation/__init__.py [hunk 1]
```diff
@@ -14,16 +14,24 @@
...
-        self.as_type = as_type
+        self.dtype = np.dtype(dtype)
         self.batch_size = batch_size
         self.preconditioner = preconditioner
 
         self.L = src.L
         self.n = src.n
 
+        if not self.src.dtype == self.basis.dtype == self.dtype:
+            # TODO: Decide if this should simply infer from src or basis.
```
Comment (janden) [inline src/aspire/estimation/__init__.py]: `src` should have precedence here.

File: src/aspire/image/__init__.py [hunk 2]
```diff
@@ -228,7 +229,7 @@ def rotate(self):
     def save(self, mrcs_filepath, overwrite=False):
         with mrcfile.new(mrcs_filepath, overwrite=overwrite) as mrc:
             # original input format (the image index first)
-            mrc.set_data(self.data.astype('float32'))
+            mrc.set_data(self.data.astype(np.float32))
```
Comment (janden) [inline src/aspire/image/__init__.py:232]: Just for my own curiosity, why is `np.float32` preferred over `'float32'`? Speed?

File: src/aspire/nufft/finufft.py [hunk 3]
```diff
@@ -5,7 +5,7 @@
 
 from aspire.nufft import Plan
 from aspire.utils import ensure
-from aspire.utils.misc import complex_type
+from aspire.utils.types import complex_type, dtype_legacy_string
```
Comment (janden) [inline src/aspire/nufft/finufft.py]: `dtype_legacy_string` is unused.

File: src/aspire/orientation/__init__.py [hunk 4]
```diff
@@ -363,6 +367,7 @@ def _get_shift_equations_approx(self, equations_factor=1, max_memory=4000):
             shift_eq[idx] = -1 * coeffs if is_pf_j_flipped else coeffs
 
         # create sparse matrix object only containing non-zero elements
+        # Note, this line is particularly sensitive to precision
```
Comment (janden) [inline src/aspire/orientation/__init__.py]: That's very strange, considering that all the arrays going into it can be `float32`. Is there some convergence issue later on that's causing problems? (According to the [docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.lsqr.html), `lsqr` seems to have a hardcoded tolerance of `1e-8`, which might be the culprit.)

File: src/aspire/source/simulation.py [hunk 5]
```diff
@@ -35,22 +35,28 @@ def __init__(self, L=8, n=1024, vols=None, states=None, filters=None,
...
             self.vols = self._gaussian_blob_vols(L=self.L, C=C, seed=seed)
         else:
             assert isinstance(vols, Volume)
             self.vols = vols
 
+        if self.vols.dtype != self.dtype:
+            logger.warning(
+                f'{self.__class__.__name__}'
+                f' vols.dtype {self.vols.dtype} != self.dtype {self.dtype}.')
+            # TODO: Decide if should just warn, raise, or convert here.
```
Comment (janden) [inline src/aspire/source/simulation.py]: Ultimately raise, but we can leave it at warn for now.

File: src/aspire/volume/__init__.py [hunk 6]
```diff
@@ -95,6 +99,13 @@ def __rmul__(self, otherL):
         return self * otherL
 
     def project(self, vol_idx, rot_matrices):
+        if rot_matrices.dtype != self.dtype:
+            logger.warning(
+                f'{self.__class__.__name__}'
+                f' rot_matrices.dtype {rot_matrices.dtype}'
+                f' != self.dtype {self.dtype}.')
+            # TODO: decide if should just warn, raise, or convert here.
```
Comment (janden) [inline src/aspire/volume/__init__.py]: Ultimately, raise. Warn for now.

File: tests/test_FBbasis2D.py [hunk 7]
```diff
@@ -66,18 +67,20 @@ def testFBBasis2DEvaluate(self):
...
+              ],
+            dtype=self.dtype
         )
         result = self.basis.evaluate(coeffs)
 
         self.assertTrue(np.allclose(
             result,
-            np.load(os.path.join(DATA_DIR, 'fbbasis_evaluation_8_8.npy')).T  #RCOPT
+            np.load(os.path.join(DATA_DIR, 'fbbasis_evaluation_8_8.npy')).T,  #RCOPT
+            atol=1e-6 if self.dtype == np.float32 else 1e-8
```
Comment (janden) [inline tests/test_FBbasis2D.py]: There's no way this is handled elegantly when both arguments to `allclose` are `float32`? If not, we should probably hardcode these tolerances in one place and call something like `test_tolerance(dtype)` when needed.

File: tests/test_batched_covar2d.py [hunk 8]
```diff
@@ -76,8 +77,15 @@ def test01(self):
         mean_bcov2d = self.bcov2d.get_mean()
         covar_bcov2d = self.bcov2d.get_covar(noise_var=noise_var)
 
-        self.assertTrue(np.allclose(mean_cov2d, mean_bcov2d))
-        self.assertTrue(self.blk_diag_allclose(covar_cov2d, covar_bcov2d))
+        self.assertTrue(np.allclose(
+            mean_cov2d,
+            mean_bcov2d,
+            atol=1e-5 if self.dtype == np.float32 else 1e-8))
```
Comment (janden) [inline tests/test_batched_covar2d.py]: Eesh, now it's `1e-5`?

File: tests/test_covar2d.py [hunk 9]
```diff
@@ -88,7 +89,7 @@ def testGetCovarCTF(self):
     def testGetCovarCTFShrink(self):
         results = np.load(os.path.join(DATA_DIR, 'clean70SRibosome_cov2d_covarctf_shrink.npy'))
         covar_opt = {'shrinker': 'frobenius_norm', 'verbose': 0, 'max_iter': 250, 'iter_callback': [],
-                     'store_iterates': False, 'rel_tolerance': 1e-12, 'precision': 'float64'}
+                     'store_iterates': False, 'rel_tolerance': 1e-12, 'precision': 'float32'}
```
Comment (janden) [inline tests/test_covar2d.py]: Not `self.dtype`?

**Review by janden — CHANGES_REQUESTED** [review]
This looks good, but there's something I don't understand: in some places we check for inconsistent dtypes (like in `FFBBasis2D`) but not in others (`PSWFBasis2D` and `BatchedRotCov2D`). Is there a specific reason for this? (For example, the inconsistency violation doesn't show up in the tests?)

File: tests/test_FBbasis2D.py (hunk 7 above)
Comment (garrettwrong) [reply tests/test_FBbasis2D.py]: 1) Negative, their default 1e-8 is already meant to be flexible for singles, and for many things it is. I think if you expect to feed certain solvers noise this expectation will not hold up.

2) I can do that but they're not all the same (well I can make them, but I would prefer them tighter). We'd have to select the crudest value (so probably 1e-5) or go back to adhocs.

File: tests/test_batched_covar2d.py (hunk 8 above)
Comment (garrettwrong) [reply tests/test_batched_covar2d.py]: Yes, I wasn't thrilled about this either, but that is what I found on the cross platform CI for our tests.

File: tests/test_covar2d.py (hunk 9 above)
Comment (garrettwrong) [reply tests/test_covar2d.py]: Yeah, looking at it now I think I'll want to go through some of the opt dictionaries. Good catch, thanks.

File: src/aspire/nufft/finufft.py (hunk 3 above)
Comment (garrettwrong) [reply src/aspire/nufft/finufft.py]: Oops, leftover from rebasing this (originally finufftv2 was implemented before this change set here, but..). Slipped through. Will rm.

File: src/aspire/image/__init__.py (hunk 2 above)
Comment (garrettwrong) [reply src/aspire/image/__init__.py:232]: Generally I always see use of the string names considered legacy. Sometimes they're just poor/ambiguous like in the case of `'complex'`.  I'd rather give np the np object.

One might also prefer using the strings only for special places in the code for other reasons...

File: tests/test_covar2d.py (hunk 9 above)
Comment (garrettwrong) [reply tests/test_covar2d.py]: I hit all the _opt dictionaries that turned up with grep to use self.dtype (source code and tests). Thanks for pointing out, I over looked it.

File: src/aspire/orientation/__init__.py (hunk 4 above)
Comment (garrettwrong) [reply src/aspire/orientation/__init__.py]: I was a little worried what we would go on to do with that matrix might be sensitive... but it is true that the results follow the same level of tolerance seen elsewhere in that test (1e-6 for singles), hopefully that will satisfy the other platforms via CI too. pushing.

I added some lines that will log out the lsqr numerical summary for DEBUG level (so to the log file and for `pytest -s` atm).

File: src/aspire/estimation/__init__.py (hunk 1 above)
Comment (garrettwrong) [reply src/aspire/estimation/__init__.py]: Cool, that simplified the places we use this from explicit to pass through (inferred from src). Good stuff.

File: tests/test_FBbasis2D.py (hunk 7 above)
Comment (garrettwrong) [reply tests/test_FBbasis2D.py]: I have implemented `utest_tolerance` at 1e-5 and replaced the inline `atol` statements.

**Comment by garrettwrong** [comment]
> This looks good, but there's something I don't understand: in some places we check for inconsistent dtypes (like in FFBBasis2D) but not in others (PSWFBasis2D and BatchedRotCov2D). Is there a specific reason for this? (For example, the inconsistency violation doesn't show up in the tests?)

Logging lines like that indicate there was a problem in the code that required debugging.  After I resolved the problem I just adjusted and left the logging lines.  I didn't go though adding logging lines as busy work.  Once at the less used classes you mention, most of the errors were resolved.  Its likely things are untested that I haven't seen yet, but that is out of scope.

That sort of raises the other reason... the ASPIRE classes should be refactored to actually use OO programming so superficial code (like these checks) are inherited instead of copy pasta everywhere. That is another can of worms entirely...  I really didn't want this to turn into another did/did-not actually develop classes/OO thing. I just want the types that hit the numerical methods to not be a free for all... and that code which admits dtypes actually works....  If it sometimes even has debugging aids, that's a perk, not what I set out to do...

If you would prefer for consistency that I add the logging lines I can do that (but I expect it would just get ripped out later anyway).

Currently the CI is failing for some trivial merge conflicts. I'll rebase it after your review so the commits don't disappear...

**Comment by garrettwrong** [comment]
Just a heads up, I rebased locally and I am finding that the recent `test06` addition has a max abs delta of `2.8E-4` vs `1.7E-11` for singles vs doubles respectively (presumably worse on other platforms).  If an order of magnitude difference in tolerance is expected/acceptable for the batch algorithm I can scale the utest_tolerance by 10 for that single test to keep this moving.

File: src/aspire/image/__init__.py (hunk 2 above)
Comment (janden) [reply src/aspire/image/__init__.py:232]: Makes sense.

File: src/aspire/orientation/__init__.py (hunk 4 above)
Comment (janden) [reply src/aspire/orientation/__init__.py]: Ok good.

File: tests/test_FBbasis2D.py (hunk 7 above)
Comment (janden) [reply tests/test_FBbasis2D.py]: If I understand correctly, `np.isclose` has two tolerances: relative (default `1e-5`) and absolute (default `1e-8`). The latter is only relevant for values close to zero (where a relative error doesn't make much sense, since it would be `1e-5` times zero). In other words, the absolute tolerance is there to make sure that machine epsilon and zero are considered “close.” Yet machine epsilon for `float32` is bigger than `1e-8` (`np.finfo(np.float32).eps` gives `1.192e-7` for me), so it's not going to catch those. For example, the following fails on my laptop:
```
x = np.array(1 / 3, dtype=np.float32) * 3 - np.array(1, dtype=np.float32)
np.isclose(x, 0)
```
(I get `x = 2.980e-08`.) Looking at the reference data in this particular test, we have values like `1e-17`, which I'm guessing is the cause for the test failure.

Long story short, I think we can feel safe setting our absolute tolerance a few orders of magnitude higher than machine epsilon. It's the relative tolerance that's important (and that stays at `1e-5`). For the exceptional cases where we have values that are essentially zero, it's ok to be a little loose.

(The flip side of this is that we can probably tighten the tolerances for `float64`. A relative error of `1e-5` and an absolute error of `1e-8` is pretty high for double precision. But that's a separate issue.)

**Comment by janden** [comment]
> Logging lines like that indicate there was a problem in the code that required debugging. After I resolved the problem I just adjusted and left the logging lines. I didn't go though adding logging lines as busy work.

Ok makes sense. Just wanted to see if I was missing anything here.

> That sort of raises the other reason... the ASPIRE classes should be refactored to actually use OO programming so superficial code (like these checks) are inherited instead of copy pasta everywhere.

Yes there's a lot of boilerplate that gets repeated that we should be able to remove (here and elsewhere). I suggest we create issues for these as we go along and get back to this after the release.

> If you would prefer for consistency that I add the logging lines I can do that (but I expect it would just get ripped out later anyway).

No let's leave them as is.

**Comment by janden** [comment]
> Just a heads up, I rebased locally and I am finding that the recent `test06` addition has a max abs delta of `2.8E-4` vs `1.7E-11` for singles vs doubles respectively (presumably worse on other platforms). If an order of magnitude difference in tolerance is expected/acceptable for the batch algorithm I can scale the utest_tolerance by 10 for that single test to keep this moving.

You're referring to https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/09eab437033bbbe6fb296d7d3301467e03b3772f/tests/test_batched_covar2d.py#L149 correct? That's a problem. There's no reason the batched covariance estimation should be that different here. Let's loosen the tolerance for now and create an issue.

File: src/aspire/denoising/denoiser_cov2d.py [hunk 10]
```diff
@@ -45,9 +45,15 @@ def denoise(self, covar_opt=None, batch_size=512):
...
-        default_opt = {'shrinker': 'frobenius_norm', 'verbose': 0, 'max_iter': 250,
-            'iter_callback': [], 'store_iterates': False, 'rel_tolerance': 1e-12,
-            'precision': 'float64'}
+        default_opt = {
+            'shrinker': 'frobenius_norm',
+            'verbose': 0,
+            'max_iter': 250,
+            'iter_callback': [],
+            'store_iterates': False,
+            'rel_tolerance': 1e-12,
```
Comment (janden) [inline src/aspire/denoising/denoiser_cov2d.py:53]: Sorry to come back to this again, but CG is unlikely to converge to `1e-12` in `float32`. Is that particular code path tested? It may be some issue with the CG implementation here. That code is overdue for a detailed examination.

**Review by janden — CHANGES_REQUESTED** [review]

File: src/aspire/denoising/denoiser_cov2d.py (hunk 10 above)
Comment (garrettwrong) [reply src/aspire/denoising/denoiser_cov2d.py:53]: The `default_*_opt` are not used directly. However, the `1e-12` value is being used in the all overrides (which are just copy pastes of the defaults, lol).  I think we'd have to print diagnostic to get into what is happening there.  If it is like other scipy methods it may bail on it's own if it thinks it reached a limit (good enough for that machine precision, for example), but that is total guess.

I've created #306 for this.

File: tests/test_FBbasis2D.py (hunk 7 above)
Comment (garrettwrong) [reply tests/test_FBbasis2D.py]: I agree about their atol and rtol. Regarding that example, you've made my point about being more careful about mixing types. `x` is a double; the arithmetic is mixed mode.  Been down that road many times, sometimes I wonder if too many :D ...

```
>>> x = np.array(1 / 3, dtype=np.float32) * 3 - np.array(1, dtype=np.float32); np.isclose(x, np.float32(0))
False
>>> x = np.array(1 / 3, dtype=np.float32) * np.float32(3) - np.array(1, dtype=np.float32); np.isclose(x, np.float32(0))
True
```
alternately
```
>>> x = np.zeros(2, dtype=np.float32)
>>> x[0] = 1 / 3
>>> x[1] = 1
>>> x[0] *= 3
>>> x[1] - x[0]
0.0
>>> np.isclose(x[1] - x[0], 0)
True
>>> (x[1] - x[0]).dtype
dtype('float32')
>>> float(x[1] - x[0]).hex()
'0x0.0p+0'
>>> ui32 = (x[1] - x[0]).view(dtype=np.uint32)
>>> ui32
0
```

I feel reasonably good about numpy's expectations in their domain, but what ASPIRE does is a little more involved (multi method) right.  Unlike some other fields, I don't have a great pulse on what the propagation of uncertainty looks like numerically in your field yet (though I still have my intuitions)... if you are comfortable with the tolerances I'm satisfied.  One sneaking intuition is that if we stay vigilant some of these can be improved, but maybe not today.

Comment (janden) [reply tests/test_FBbasis2D.py]: Huh… I didn't realize it's quite that bad. (So it's good that we're getting away from this stuff.) I'm trying to understand what's happening in that particular example for things to come out to exactly zero, but my familiarity with IEEE 754 arithmetic is not too good… Perhaps we can discuss at some point. My point is there's no reason for errors due to non-associativity to be smaller than `1e-8` in single precision, so it seems odd to have that as a default absolute tolerance.

At any rate, this is fine for now

File: src/aspire/denoising/denoiser_cov2d.py (hunk 10 above)
Comment (janden) [reply src/aspire/denoising/denoiser_cov2d.py:53]: Ok thanks.

**Comment by garrettwrong** [comment]
Pushed the rebase now that most conversations resolved. The conflicts were piling up, they should be all sorted now.

Regarding `test06` it looks like I need to bring it down even farther to pass CI platforms (though I am still waiting to see if this one can succeed).  That is highly suspect. Also a test that was previously passing `test02` now also required minor adjustment for the CI platforms. I wonder if they are related. (by now I mean, after I rebased with the changes brought into develop over the week prior).

I created #307 so we can keep moving. We can talk in the meeting soon, thanks.

**Review by janden — APPROVED** [review]
